src/include/sof/audio/buffer.h

```diff
@@ -271,6 +271,64 @@ static inline uint32_t buffer_avail_frames(struct comp_buffer *source,
 	return MIN(src_frames, sink_frames);
 }
 
+/**
+ * Returns the contiguous memory region which is available in this buffer.
+ * @param buf Component buffer
+ * @param[out] r_ptr Beginning of readable memory region
+ * @param[out] r_size Size (in bytes) of readable memory region
+ * @return 0 on success, -EINVAL if buf is NULL or both output parameters are
+ *         NULL
+ */
+static inline int buffer_contig_avail(struct comp_buffer *buf,
+	void **r_ptr, size_t *r_size)
+{
+	char *read;
+	size_t read_size;
+
+	if (!buf)
+		return -EINVAL;
+	if (!r_ptr && !r_size)
+		return -EINVAL;
+	read = buf->r_ptr;
+	read_size = buf->avail;
+	if (read + read_size > (char *)buf->end_addr)
+		read_size = (char *)buf->end_addr - read;
+	if (r_ptr)
+		*r_ptr = read;
+	if (r_size)
+		*r_size = read_size;
+	return 0;
+}
+
+/**
+ * Returns the contiguous memory region which is free in this buffer.
+ * @param buf Component buffer
+ * @param[out] w_ptr Beginning of writeable memory region
+ * @param[out] w_size Size (in bytes) of writeable memory region
+ * @return 0 on success, -EINVAL if buf is NULL or both output parameters are
+ *         NULL
+ */
+static inline int buffer_contig_free(struct comp_buffer *buf,
```

Comment:
buffer_config_x() does not match the use case. buffer_get_contiguous_free() and buffer_get_contiguous_avail() match the use case.

Reply:
Btw, how do we deal with wrap when we have 2 contiguous areas avail (and say the larger area is not reported here).

Reply:
https://github.com/thesofproject/sof/pull/2184#discussion_r356040815

Here you can see a use case (closed source component which cannot be changed and expects contiguous buffer). So dealing with wrap is done by repeatedly consuming/producing until one buffer is either empty or full (that would be up to 3 or 4 iterations).

As for your first comment, you mean renaming these functions to "buffer_get_contiguous_free"/avail?

Reply:
Yes, renaming to match usage (since we are not really configuring the buffer, but just getting an attribute).

Reply:
Yeah, anyway it just said "contig" not "config". Renaming, will also rebase while I'm at it.